module/executiondatasync/tracker/storage.go

```diff
@@ -0,0 +1,520 @@
+	globalStateKey[0] = prefixGlobalState
+	globalStateKey[1] = state
+	return globalStateKey
+}
+
+const blobRecordKeyLength = 1 + 8 + blobs.CidLength
+
+func makeBlobRecordKey(blockHeight uint64, c cid.Cid) []byte {
+	blobRecordKey := make([]byte, blobRecordKeyLength)
+	blobRecordKey[0] = prefixBlobRecord
+	binary.LittleEndian.PutUint64(blobRecordKey[1:], blockHeight)
+	copy(blobRecordKey[1+8:], c.Bytes())
+	return blobRecordKey
+}
+
+func parseBlobRecordKey(key []byte) (uint64, cid.Cid, error) {
+	blockHeight := binary.LittleEndian.Uint64(key[1:])
+	c, err := cid.Cast(key[1+8:])
+	return blockHeight, c, err
+}
+
+const latestHeightKeyLength = 1 + blobs.CidLength
+
+func makeLatestHeightKey(c cid.Cid) []byte {
+	latestHeightKey := make([]byte, latestHeightKeyLength)
+	latestHeightKey[0] = prefixLatestHeight
+	copy(latestHeightKey[1:], c.Bytes())
+	return latestHeightKey
+}
+
+func makeUint64Value(v uint64) []byte {
+	value := make([]byte, 8)
+	binary.LittleEndian.PutUint64(value, v)
+	return value
+}
+
+func getUint64Value(item *badger.Item) (uint64, error) {
+	value, err := item.ValueCopy(nil)
+	if err != nil {
+		return 0, err
+	}
+
+	return binary.LittleEndian.Uint64(value), nil
+}
+
+// getBatchItemCountLimit returns the maximum number of items that can be included in a single batch
+// transaction based on the number / total size of updates per item.
+func getBatchItemCountLimit(db *badger.DB, writeCountPerItem int64, writeSizePerItem int64) int {
+	totalSizePerItem := 2*writeCountPerItem + writeSizePerItem // 2 bytes per entry for user and internal meta
+	maxItemCountByWriteCount := db.MaxBatchCount() / writeCountPerItem
+	maxItemCountByWriteSize := db.MaxBatchSize() / totalSizePerItem
+
+	if maxItemCountByWriteCount < maxItemCountByWriteSize {
+		return int(maxItemCountByWriteCount)
+	} else {
+		return int(maxItemCountByWriteSize)
+	}
+}
+
+// TrackBlobsFun is passed to the UpdateFn provided to Storage.Update,
+// and can be called to track a list of cids at a given block height.
+// It returns an error if the update failed.
+type TrackBlobsFn func(blockHeight uint64, cids ...cid.Cid) error
+
+// UpdateFn is implemented by the user and passed to Storage.Update,
+// which ensures that it will never be run concurrently with any call
+// to Storage.Prune.
+// Any returned error will be returned from the surrounding call to Storage.Update.
+// The function must never make any calls to the Storage interface itself,
+// and should instead only modify the storage via the provided TrackBlobsFn.
+type UpdateFn func(TrackBlobsFn) error
+
+// PruneCallback is a function which can be provided by the user which
+// is called for each CID when the last height at which that CID appears
+// is pruned.
+// Any returned error will be returned from the surrounding call to Storage.Prune.
+// The prune callback can be used to delete the corresponding
+// blob data from the blob store.
+type PruneCallback func(cid.Cid) error
+
+type Storage interface {
```

Comment:
For each method here, could you note any expected error returns. If there are none, could you add the one-liner
```
No errors are expected during normal operation
```

☝️ this helps to differentiate between:
* the function doesn't have any expected error types
* the function hasn't been updated to document it's expected error types yet